classes/solid/solid_object_mirrorable.gd

```diff
@@ -0,0 +1,90 @@
+class_name SolidObjectMirrorable
+extends SolidObject
+## A StaticBody2D which can be both disabled or mirrored in the level designer.
+
+static var mirrored_shapes: Dictionary = {}
+static var unmirrored_shapes: Dictionary = {}
+
+@export var mirror = false: set = set_mirror
+
+
+func set_mirror(val):
+	# Store whether the new value is DIFFERENT from the old.
+	# This allows several things to flip correctly without storing their
+	# old positions.
+	var changed = val != mirror
+	
+	# Write to the mirroring value itself.
+	mirror = val
+	
+	# If mirroring didn't change, no child nodes need to be processed. Abort.
+	if !changed:
+		return
+	
+	# If mirroring did change, flip all children.
+	for child in get_children():
+		# Flip any child's transformation.
+		if child is Node2D:
+			(child as Node2D).position.x *= -1
+			(child as Node2D).rotation *= -1
+		
+		# Type-specific mirroring logic may also be needed.
+		
+		# If the child has its own mirroring logic, fall through into that.
+		if child.has_method("set_mirror"):
+			child.set_mirror(mirror)
+		
+		# Flip sprites.
+		elif child is Sprite2D or child is AnimatedSprite2D:
+			child.flip_h = !child.flip_h if changed else child.flip_h
+		
+		# Flip polygons and collision polygons, which conveniently share an
+		# interface as of Godot 4.1.
+		elif child is Polygon2D or child is CollisionPolygon2D:
+			child.polygon = flip_points(child.polygon)
+		
+		# Flip collision shapes that can flip.
+		elif child is CollisionShape2D:
+			var shape := (child as CollisionShape2D).shape
+			if changed and mirror:
+				# On becoming mirrored, make a copy of the shape resource
+				# so we can mirror its data without messing up any other users
+				# of the shape resource (yes, the shapes are resources!).
+				shape = shape.duplicate() # Make and store the duplicate...
```

Comment:
I assume Godot will garbage-collect shapes that are no longer in use, to avoid a memory leak?

Reply:
Yeah, Resource inherits RefCounted. I actually put a whole paragraph about that below the assignments...maybe I should've put it above and accepted the daunting-ness of it :/

Reply:
Oh, guh, I see that now. Thanks.

I think it's fine where it is, I just tuned it out because GitHub's web UI wraps the text in a weird way that makes it hard for me to pay attention to it.